### PR TITLE
WT-9805 Save the updates need to be deleted from history store and delete them later

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

File: src/include/reconcile.h [hunk 1]
```diff
@@ -227,6 +227,14 @@ struct __wt_reconcile {
     size_t supd_allocated;
     size_t supd_memsize; /* Size of saved update structures */
 
+    /*
+     * History store update to be deleted list. While reviewing updates for each page, we save
+     * WT_HS_DELETE_UPD lists here, and then delete them after we have built the disk image.
+     */
+    WT_DELETE_HS_UPD *delete_hs_upd; /* Updates to delete from history store */
+    uint32_t delete_hs_upd_next;
```
Comment (etienneptl) [inline src/include/reconcile.h:235]: Should we call it `delete_hs_upd_size` or `delete_hs_upd_count`?

File: src/history/hs_rec.c [hunk 2]
```diff
@@ -1046,3 +1046,103 @@ __hs_delete_reinsert_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, ui
...
+    WT_DECL_RET;
+    bool hs_read_committed;
+#ifdef HAVE_DIAGNOSTIC
+    WT_TIME_WINDOW *hs_tw;
+#endif
+
+    if (r->hs_cursor == NULL)
+        WT_RET(__wt_curhs_open(session, NULL, &r->hs_cursor));
+    hs_read_committed = F_ISSET(r->hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
+    F_SET(r->hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
```
Comment (etienneptl) [inline src/history/hs_rec.c:1067]: Can we add this comment:
```
/* Ensure we can see all the content in the history store. */
```

File: src/history/hs_rec.c [hunk 3]
```diff
@@ -1046,3 +1046,103 @@ __hs_delete_reinsert_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, ui
...
+
+    /* No need to delete from the history store if it is already obsolete. */
+    if (tombstone != NULL && __wt_txn_upd_visible_all(session, tombstone))
+        goto done;
+
+    r->hs_cursor->set_key(r->hs_cursor, 4, S2BT(session)->id, key, WT_TS_MAX, UINT64_MAX);
+    WT_ERR_NOTFOUND_OK(__wt_curhs_search_near_before(session, r->hs_cursor), true);
+    /* It's possible the value in the history store becomes obsolete concurrently. */
+    if (ret == WT_NOTFOUND) {
+        WT_ASSERT(session, tombstone != NULL && __wt_txn_upd_visible_all(session, tombstone));
```
Comment (etienneptl) [inline src/history/hs_rec.c:1077]: Instead of the `WT_ASSERT`, should we do the same `if` as on line 1070? This race condition seems to be a problem though. If we have `ret == WT_NOTFOUND` but more content has been inserted to the hs just then, the assert will be false?

File: src/history/hs_rec.c [hunk 4]
```diff
@@ -1046,3 +1046,103 @@ __hs_delete_reinsert_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, ui
...
+
+    WT_ERR(r->hs_cursor->remove(r->hs_cursor));
+done:
+    if (tombstone != NULL)
+        F_CLR(tombstone, WT_UPDATE_TO_DELETE_FROM_HS | WT_UPDATE_HS);
+    F_CLR(upd, WT_UPDATE_TO_DELETE_FROM_HS | WT_UPDATE_HS);
+
+err:
+    if (!hs_read_committed)
+        F_CLR(r->hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
```
Comment (etienneptl) [inline src/history/hs_rec.c:1104]: In the case of an error, we would jump to the label `err` and clear the flag of `hs_cursor` if required. Why wouldn't we clear the flag(s) for the `tombstone` and `update` as well?
Instead of:
```
WT_ERR(r->hs_cursor->remove(r->hs_cursor));
```
We could simply do:
```
ret = r->hs_cursor->remove(r->hs_cursor);
```
What do you think?

File: src/include/btmem.h [hunk 5]
```diff
@@ -246,6 +246,17 @@ struct __wt_save_upd {
     bool restore; /* Whether to restore this saved update chain */
 };
 
+/*
+ * WT_DELETE_HS_UPD --
+ *	Update needs to be deleted from the history store.
```
Comment (etienneptl) [inline src/include/btmem.h:251]: I would say "Update **that** needs ...". Or "Information related to an update that should be removed from the history store".

File: src/include/reconcile.h [hunk 6]
```diff
@@ -227,6 +227,14 @@ struct __wt_reconcile {
     size_t supd_allocated;
     size_t supd_memsize; /* Size of saved update structures */
 
+    /*
+     * History store update to be deleted list. While reviewing updates for each page, we save
```
Comment (etienneptl) [inline src/include/reconcile.h:231]: I would reword with:
"List of history store updates to be deleted. While reviewing updates for each page, we save them here and delete them after we have built the disk image.".

File: src/reconcile/rec_visibility.c [hunk 7]
```diff
@@ -39,6 +39,27 @@ __rec_update_save(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, WT_
     return (0);
 }
 
+/*
+ * __rec_delete_hs_upd_save --
+ *     Save a WT_DELETE_HS_UPDATE list to delete the update from the history store later.
```
Comment (etienneptl) [inline src/reconcile/rec_visibility.c:44]: `WT_DELETE_HS_UPDATE` -> `WT_DELETE_HS_UPD`

Comment (etienneptl) [reply src/reconcile/rec_visibility.c:44]: Could we say:
"Save an update into a WT_DELETE_HS_UPDATE list"? I am not sure if it's necessary to say that it is an update that needs to be deleted from the HS later. This function only serves the purpose of adding some content to a `WT_DELETE_HS_UPD`.

File: src/history/hs_rec.c [hunk 8]
```diff
@@ -1046,3 +1046,103 @@ __hs_delete_reinsert_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, ui
 
     return (ret);
 }
+
+/*
+ * __hs_delete_record --
+ *     Delete the update left in the history store
```
Comment (etienneptl) [inline src/history/hs_rec.c:1052]: Can we reword to:
"Delete an update from the history store"? We can probably add information about the case when we don't have to delete anything.

File: src/history/hs_rec.c [hunk 9]
```diff
@@ -1046,3 +1046,103 @@ __hs_delete_reinsert_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, ui
...
+int
+__wt_hs_delete_updates(WT_SESSION_IMPL *session, WT_RECONCILE *r)
+{
+    WT_DECL_ITEM(key);
+    WT_DECL_RET;
+    WT_DELETE_HS_UPD *delete_hs_upd;
+    uint32_t i;
+    uint8_t *p;
+
+    WT_ERR(__wt_scr_alloc(session, WT_INTPACK64_MAXSIZE, &key));
```
Comment (etienneptl) [inline src/history/hs_rec.c:1121]: Do you think we should check for the number of elements to be deleted before making this call?
it would be something like this:
```
if(r->delete_hs_upd_next == 0) goto done;
WT_ERR(__wt_scr_alloc(session, WT_INTPACK64_MAXSIZE, &key));
```

File: src/include/reconcile.h (hunk 1 above)
Comment (quchenhao) [reply src/include/reconcile.h:235]: The name is consistent to other similar use cases. I intend to keep the same pattern.

File: src/history/hs_rec.c (hunk 3 above)
Comment (quchenhao) [reply src/history/hs_rec.c:1077]: No, we already have a cursor position, inserting more things will not impact the current cursor position.

File: src/history/hs_rec.c (hunk 4 above)
Comment (quchenhao) [reply src/history/hs_rec.c:1104]: If we failed to remove, then the update is still in the history store, we cannot clear the flag in this case. We should retry to remove again next time.

File: src/history/hs_rec.c (hunk 9 above)
Comment (quchenhao) [reply src/history/hs_rec.c:1121]: yes, we can do that. What we gain is not allocating the key buffer in this case.

File: src/history/hs_rec.c (hunk 3 above)
Comment (quchenhao) [reply src/history/hs_rec.c:1077]: No, it cannot happen, the assert is on tombstone. Doesn't have anything to do with newly inserted content.

File: src/reconcile/rec_visibility.c (hunk 7 above)
Comment (quchenhao) [reply src/reconcile/rec_visibility.c:44]: I would still keep the history store here to make the intention clear.

**Review by etienneptl — APPROVED** [review]
LGTM!
Good job @quchenhao, would be great to update de description section of the PR though with a quick summary of the changes.

File: src/history/hs_rec.c [hunk 10]
```diff
@@ -60,65 +60,6 @@ __hs_verbose_cache_stats(WT_SESSION_IMPL *session, WT_BTREE *btree)
         cache->hs_verb_gen_write = ckpt_gen_current;
 }
 
-/*
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/history/hs_rec.c:63]: This is still in this file but I moved it to the end of the file.

File: src/history/hs_rec.c [hunk 11]
```diff
@@ -441,44 +393,6 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_MULTI *mult
         enable_reverse_modify =
           (WT_STREQ(btree->value_format, "S") || WT_STREQ(btree->value_format, "u"));
 
-        /*
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/history/hs_rec.c:444]: yes, that's correct.

File: src/include/btmem.h [hunk 12]
```diff
@@ -246,6 +246,17 @@ struct __wt_save_upd {
...
+ */
+struct __wt_delete_hs_upd {
+    WT_INSERT *ins; /* Insert list reference */
+    WT_ROW *rip;    /* Original on-page reference */
+    WT_UPDATE *upd;
+    WT_UPDATE *tombstone;
+};
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/include/btmem.h:258]: I put it here because __wt_save_upd is here. May be we should move both of them to reoncile.h

[the remark this answers is not in the document]
Comment (quchenhao) [reply src/include/btmem.h:258]: __wt_save_upd is referred in multi block so it is here. I think we can move __wt_delete_hs_upd to reconcile.h

File: src/btree/bt_cursor.c [hunk 13]
```diff
@@ -612,6 +612,8 @@ __wt_btcur_reset(WT_CURSOR_BTREE *cbt)
     WT_STAT_CONN_DATA_INCR(session, cursor_reset);
 
     F_CLR(cursor, WT_CURSTD_KEY_SET | WT_CURSTD_VALUE_SET);
+    cbt->upd_value->type = WT_UPDATE_INVALID;
+    WT_TIME_WINDOW_INIT(&cbt->upd_value->tw);
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/btree/bt_cursor.c:617]: The code is OK to not clear it. There is no correctness issue. I did this because I see in the core dump that we have garbage value for the values so I think better to fix that to not confuse us when debugging.

File: src/history/hs_rec.c [hunk 14]
```diff
@@ -407,6 +347,19 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_MULTI *mult
         if (list->onpage_upd == NULL)
             continue;
 
+        /* Skip aborted updates. */
+        for (upd = list->onpage_upd->next; upd != NULL && upd->txnid == WT_TXN_ABORTED;
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/history/hs_rec.c:352]: done

File: src/include/btmem.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/include/btmem.h]: done

File: src/history/hs_rec.c [hunk 15]
```diff
@@ -1132,3 +1047,108 @@ __hs_delete_reinsert_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, ui
...
+    WT_DECL_RET;
+    WT_DELETE_HS_UPD *delete_hs_upd;
+    uint32_t i;
+    uint8_t *p;
+
+    /* Nothing to delete from the history store. */
+    if (r->delete_hs_upd == NULL)
+        return (0);
+
+    WT_ERR(__wt_scr_alloc(session, WT_INTPACK64_MAXSIZE, &key));
```
Comment (agorrod) [inline src/history/hs_rec.c:1127]: I think this can use WT_RET - if it returns an error, the scratch buffer won't have been allocated, so doesn't need to be cleaned up.

File: src/history/hs_rec.c [hunk 16]
```diff
@@ -1132,3 +1047,108 @@ __hs_delete_reinsert_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, ui
...
+
+    for (delete_hs_upd = r->delete_hs_upd, i = 0; i < r->delete_hs_upd_next; ++delete_hs_upd, ++i) {
+        switch (r->page->type) {
+        case WT_PAGE_COL_FIX:
+        case WT_PAGE_COL_VAR:
+            p = key->mem;
+            WT_ERR(__wt_vpack_uint(&p, 0, WT_INSERT_RECNO(delete_hs_upd->ins)));
+            key->size = WT_PTRDIFF(p, key->data);
```
Comment (agorrod) [inline src/history/hs_rec.c:1135]: Did you base this code off other code? I can't see it being handled prior to this change - was there just a bug in how column store content was cleaned up?
Could you get around calculating it by saving a recno in the delete_hs_upd structure?

Comment (quchenhao) [reply src/history/hs_rec.c:1135]: I referred to the code in __wt_hs_insert_updates. By saving ins and rip we can pack the key for all the cases for column store and row store. If we choose to save recno, we still need to save ins and rip as there is case that row store needs ins if it only has an insert list.

File: src/history/hs_rec.c [hunk 17]
```diff
@@ -1132,3 +1047,108 @@ __hs_delete_reinsert_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, ui
...
+    if (tombstone != NULL && __wt_txn_upd_visible_all(session, tombstone))
+        goto done;
+
+    r->hs_cursor->set_key(r->hs_cursor, 4, S2BT(session)->id, key, WT_TS_MAX, UINT64_MAX);
+    WT_ERR_NOTFOUND_OK(__wt_curhs_search_near_before(session, r->hs_cursor), true);
+    /* It's possible the value in the history store becomes obsolete concurrently. */
+    if (ret == WT_NOTFOUND) {
+        WT_ASSERT(session, tombstone != NULL && __wt_txn_upd_visible_all(session, tombstone));
+        ret = 0;
+        goto done;
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/history/hs_rec.c]: done

File: src/history/hs_rec.c [hunk 18]
```diff
@@ -1132,3 +1047,107 @@ __hs_delete_reinsert_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, ui
...
+        switch (r->page->type) {
+        case WT_PAGE_COL_FIX:
+        case WT_PAGE_COL_VAR:
+            p = key->mem;
+            WT_ERR(__wt_vpack_uint(&p, 0, WT_INSERT_RECNO(delete_hs_upd->ins)));
+            key->size = WT_PTRDIFF(p, key->data);
+            break;
+        case WT_PAGE_ROW_LEAF:
+            if (delete_hs_upd->ins == NULL) {
+                WT_ERR(__wt_row_leaf_key(session, r->page, delete_hs_upd->rip, key, false));
```
Comment (agorrod) [inline src/history/hs_rec.c:1138]: The code this came from was using `WT_WITH_BTREE` - should this code do that as well? Is it possible that without the right btree, the resolution will use the history store tree?

File: src/history/hs_rec.c (hunk 16 above)
Comment (agorrod) [reply src/history/hs_rec.c:1134]: Got it - I'd rather not have two copies of this code. It doesn't look like it would be obvious to factor it out into a shared function, but could you take a look?

File: src/history/hs_rec.c (hunk 18 above)
Comment (quchenhao) [reply src/history/hs_rec.c:1138]: Good point, I'll fix that.

Comment (quchenhao) [reply src/history/hs_rec.c]: done

**Review by agorrod — APPROVED** [review]
Thanks @quchenhao - I'm happy with this now. It definitely requires a full review from someone else as well.

I've added in stress testing to the PR - I think w should wait for that to complete before merging.

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Comment by quchenhao** [comment]
evergreen merge
